### PR TITLE
new_audit(dobetterweb): add new audit for doctype

### LINKED ISSUE #5213: Audit doctype
Recommend `<!DOCTYPE html>` to avoid quirks mode.

https://developer.mozilla.org/en-US/docs/Glossary/Doctype

### PR DESCRIPTION
**Pull request by schalkneethling**
[WIP] Adds audit for `Doctype`

Closes #5213

### CODE REVIEW

File: lighthouse-core/gather/gatherers/dobetterweb/doctype.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+class Doctype extends Gatherer {
+  /**
+   * @param {LH.Gatherer.PassContext} passContext
+   * @return {Promise<LH.Artifacts['Doctype']>}
+   */
+  afterPass(passContext) {
+    const driver = passContext.driver;
+
+    return driver.getDocument().then(node => node);
```
Comment (schalkneethling) [inline lighthouse-core/gather/gatherers/dobetterweb/doctype.js]: I am hoping this will return `document` as it does not look like I can call `driver.document` as what I am really after is `document` and then from that `document.name`

File: lighthouse-core/audits/dobetterweb/doctype.js [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    }
+
+    if (artifacts.Doctype.name.trim().length === 0) {
+      return {
+        rawValue: false,
+        explanation: 'Doctype name cannot be empty',
+      };
+    }
+
+    return {
```
Comment (paulirish) [inline lighthouse-core/audits/dobetterweb/doctype.js]: over here we'll assert that `.name === 'html'` 

link up https://html.spec.whatwg.org/multipage/parsing.html#the-initial-insertion-mode in a comment

File: lighthouse-core/gather/gatherers/dobetterweb/doctype.js (hunk 1 above)
Comment (paulirish) [reply lighthouse-core/gather/gatherers/dobetterweb/doctype.js]: rather than using the protocol's DOM methods, let's just grab the doctype via javascript.

define a function declaration in the gatherer.. something like 

```js
function getDoctype() {
  const {name, publicId, systemId} = document.doctype;
  return {name, publicId, systemId};
}
```

( i think we should grab the name, publicId, and systemId )

then, on L18 you can do 
```js
driver.evaluateAsync(`(${getDoctype.toString()}())`)
```

see the cache-contents gatherer for a model.

File: lighthouse-core/audits/dobetterweb/doctype.js (hunk 2 above)
Comment (paulirish) [reply lighthouse-core/audits/dobetterweb/doctype.js]: and i guess that publicId and systemId are blank (if i'm reading the spec right)

File: lighthouse-core/audits/dobetterweb/doctype.js [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+class Doctype extends Audit {
+  /**
+   * @return {LH.Audit.Meta}
+   */
+  static get meta() {
+    return {
+      name: 'doctype',
+      description: 'Document uses HTML5 doctype',
+      failureDescription: 'Document does not use HTML5 doctype or no doctype found',
```
Comment (paulirish) [inline lighthouse-core/audits/dobetterweb/doctype.js]: > Page is missing the HTML doctype

File: lighthouse-core/audits/dobetterweb/doctype.js [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+const Audit = require('../audit');
+
+class Doctype extends Audit {
+  /**
+   * @return {LH.Audit.Meta}
+   */
+  static get meta() {
+    return {
+      name: 'doctype',
+      description: 'Document uses HTML5 doctype',
```
Comment (paulirish) [inline lighthouse-core/audits/dobetterweb/doctype.js]: > Page has the HTML doctype

**Comment by schalkneethling** [comment]
Great stuff, thanks for the review and feedback @paulirish - I will work on this next round and update the PR.

**Comment by paulirish** [comment]
Cool! btw here's how to add to the config (and temporarily only run this one audit, skipping the rest)

```diff
diff --git a/lighthouse-core/config/constants.js b/lighthouse-core/config/constants.js
index a8995d63..e9808423 100644
--- a/lighthouse-core/config/constants.js
+++ b/lighthouse-core/config/constants.js
@@ -42,7 +42,9 @@ const defaultSettings = {
   blockedUrlPatterns: null,
   additionalTraceCategories: null,
   extraHeaders: null,
-  onlyAudits: null,
+  onlyAudits: [
+    'doctype',
+  ],
   onlyCategories: null,
   skipAudits: null,
 };
diff --git a/lighthouse-core/config/default-config.js b/lighthouse-core/config/default-config.js
index f06ab1f5..cbb0b6b1 100644
--- a/lighthouse-core/config/default-config.js
+++ b/lighthouse-core/config/default-config.js
@@ -32,6 +32,7 @@ module.exports = {
       'dobetterweb/all-event-listeners',
       'dobetterweb/anchors-with-no-rel-noopener',
       'dobetterweb/appcache',
+      'dobetterweb/doctype',
       'dobetterweb/domstats',
       'dobetterweb/js-libraries',
       'dobetterweb/optimized-images',
@@ -162,6 +163,7 @@ module.exports = {
     'byte-efficiency/uses-responsive-images',
     'byte-efficiency/efficient-animated-content',
     'dobetterweb/appcache-manifest',
+    'dobetterweb/doctype',
     'dobetterweb/dom-size',
     'dobetterweb/external-anchors-use-rel-noopener',
     'dobetterweb/geolocation-on-start',
@@ -371,6 +373,7 @@ module.exports = {
         {id: 'no-document-write', weight: 1},
         {id: 'external-anchors-use-rel-noopener', weight: 1},
         {id: 'geolocation-on-start', weight: 1},
+        {id: 'doctype', weight: 1},
         {id: 'no-vulnerable-libraries', weight: 1},
         {id: 'notification-on-start', weight: 1},
         {id: 'deprecations', weight: 1},
```



and to then iterate quickly you can do

```sh
# just run gatherers
lighthouse http://example.com -G

# just run audits from previously gathered artifacts
lighthouse http://example.com -A

# regular complete run but save artifacts to disk
lighthouse http://example.com -GA
```

**Comment by schalkneethling** [comment]
Sorry folks, have been down with a side of the swine flu. Update coming soon.

**Comment by schalkneethling** [comment]
Alrighty, finally got around to this again. Thanks so much for all of the feedback and guidance. It seems to me to be working so yay! \o/

screen shot 2018-06-17 at 12 59 59

With that said, running `yarn test` throws up a bunch of errors which I am finding a hard time correlating to the work I have done here. Let me know once you have had a chance to look over the changes. Thanks again.

File: lighthouse-core/audits/dobetterweb/doctype.js [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+      requiredArtifacts: ['Doctype'],
+    };
+  }
+
+  /**
+   * @param {LH.Artifacts} artifacts
+   * @return {LH.Audit.Product}
+   */
+  static audit(artifacts) {
+    if (artifacts.Doctype === null) {
```
Comment (ev1stensberg) [inline lighthouse-core/audits/dobetterweb/doctype.js]: Just a nitpick here, you can do `if(!artifacts.Doctype)`, and maybe the naming should be lower case?

File: lighthouse-core/audits/dobetterweb/doctype.js [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+      return {
+        rawValue: false,
+        explanation: 'Doctype name cannot be empty',
+      };
+    }
+
+    /* Note that the value for name is case sensitive,
+       and must be the string `html`. For details see:
+       https://html.spec.whatwg.org/multipage/parsing.html#the-initial-insertion-mode */
+    if (artifacts.Doctype.name.trim() === 'html') {
```
Comment (ev1stensberg) [inline lighthouse-core/audits/dobetterweb/doctype.js]: cleaner to store this in a variable?

File: lighthouse-core/audits/dobetterweb/doctype.js [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    if (artifacts.Doctype.name.trim().length === 0) {
+      return {
+        rawValue: false,
+        explanation: 'Doctype name cannot be empty',
+      };
+    }
+
+    /* Note that the value for name is case sensitive,
```
Comment (ev1stensberg) [inline lighthouse-core/audits/dobetterweb/doctype.js]: Just mention it's case sensitive and a link like: `case sensitive: https://foo.bar`

File: lighthouse-core/audits/dobetterweb/doctype.js [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+       and must be the string `html`. For details see:
+       https://html.spec.whatwg.org/multipage/parsing.html#the-initial-insertion-mode */
+    if (artifacts.Doctype.name.trim() === 'html') {
+      return {
+        rawValue: true,
+      };
+    } else {
+      return {
+        rawValue: false,
+        explanation: 'Doctype name must be the lowercase string html',
```
Comment (ev1stensberg) [inline lighthouse-core/audits/dobetterweb/doctype.js]: discard `the`?

File: lighthouse-core/gather/gatherers/dobetterweb/doctype.js [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+
+class Doctype extends Gatherer {
+  /**
+   * @param {LH.Gatherer.PassContext} passContext
+   * @return {Promise<LH.Artifacts['Doctype']>}
+   */
+  afterPass(passContext) {
+    const driver = passContext.driver;
+    /** @type {Object<string>} */
+    return driver.evaluateAsync(`(${getDoctype.toString()}())`);
```
Comment (ev1stensberg) [inline lighthouse-core/gather/gatherers/dobetterweb/doctype.js]: Maybe extract this to a var? A lot of encapsulation

File: lighthouse-core/test/audits/dobetterweb/doctype-test.js [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+
+const Audit = require('../../../audits/dobetterweb/doctype.js');
+const assert = require('assert');
+
+/* eslint-env mocha */
+
+describe('DOBETTERWEB: doctype audit', () => {
+  it('fails when HTML does not contain a doctype', () => {
+    const auditResult = Audit.audit({
+      Doctype: null,
```
Comment (ev1stensberg) [inline lighthouse-core/test/audits/dobetterweb/doctype-test.js]: See `lowercase` suggestion

File: package.json [hunk 11]
```diff
@@ -23,9 +23,7 @@
     "build-viewer": "cd ./lighthouse-viewer && yarn build",
     "clean": "rimraf *.report.html *.report.dom.html *.report.json *.screenshots.html *.screenshots.json *.devtoolslog.json *.trace.json || true",
     "lint": "[ \"$CI\" = true ] && eslint --quiet -f codeframe . || eslint .",
-
     "smoke": "node lighthouse-cli/test/smokehouse/run-smoke.js",
```
Comment (ev1stensberg) [inline package.json]: Don't know if these whitespace trims are needed

**Comment by brendankenny** [comment]
It's kind of buried in the end of the `yarn diff:sample-json` test output, but you need to run `yarn update:sample-json` to update the "golden lhr" expectations file since there will be a new audit output in there.

The Viewer test uses that file for expected output as well, so running that command will fix that test too.

File: lighthouse-core/audits/dobetterweb/doctype.js (hunk 5 above)
Comment (brendankenny) [reply lighthouse-core/audits/dobetterweb/doctype.js]: > and maybe the naming should be lower case?

It's uppercase for historical reasons (it automatically uses the the gatherer class name), and we have wanted to change that at some point, but not starting with this PR :)

File: lighthouse-core/gather/gatherers/dobetterweb/doctype.js [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+const Gatherer = require('../gatherer');
+
+/**
+ * Get and return `name`, `publicId`, `systemId` from
+ * `document.doctype`
+ * @return {Object<string>}
```
Comment (brendankenny) [inline lighthouse-core/gather/gatherers/dobetterweb/doctype.js]: I believe  for the typescript compiler, `Object` needs the key and value types, so it's `Object<string, string>`, though it's only complaining for one of these, so I'm not positive.

Regardless, it would probably be better to do the more specific `@return {{name: string, publicId: string, systemId: string}}` anyways.

File: lighthouse-core/gather/gatherers/dobetterweb/doctype.js [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+}
+
+class Doctype extends Gatherer {
+  /**
+   * @param {LH.Gatherer.PassContext} passContext
+   * @return {Promise<LH.Artifacts['Doctype']>}
+   */
+  afterPass(passContext) {
+    const driver = passContext.driver;
+    /** @type {Object<string>} */
```
Comment (brendankenny) [inline lighthouse-core/gather/gatherers/dobetterweb/doctype.js]: `@type {{name: string, publicId: string, systemId: string}}`

File: package-lock.json [hunk 14]
```diff
@@ -0,0 +1,11762 @@
+{
```
Comment (brendankenny) [inline package-lock.json]: remove from PR?

File: package.json (hunk 11 above)
Comment (brendankenny) [reply package.json]: yeah, these are (annoyingly) trimmed automatically when running one of several different yarn/npm commands, but since no functional changes to the file can just remove from the PR

File: lighthouse-core/audits/dobetterweb/doctype.js (hunk 5 above)
Comment (schalkneethling) [reply lighthouse-core/audits/dobetterweb/doctype.js]: @ev1stensberg I totally could but, I tend to want to be explicit. And as the compiler/minifier will optimise the code in the end, it seems ok. Thoughts?

File: lighthouse-core/audits/dobetterweb/doctype.js [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+      };
+    }
+  }
```
Comment (brendankenny) [inline lighthouse-core/audits/dobetterweb/doctype.js]: I think @paulirish also wanted to give audit feedback on the "`publicId` and `systemId` should be empty string cases" to get full coverage of the spec? Though is that only xhtml or how does that even happen :)

File: lighthouse-core/test/audits/dobetterweb/doctype-test.js [hunk 16]
```diff
@@ -0,0 +1,20 @@
...
+    });
+    assert.equal(auditResult.rawValue, false);
+  });
```
Comment (brendankenny) [inline lighthouse-core/test/audits/dobetterweb/doctype-test.js]: also want to test the empty string, non-"html", and "html" success cases (as well as `publicId` and `systemId` if added?)

File: package.json (hunk 11 above)
Comment (schalkneethling) [reply package.json]: This was required because there was a conflict reported in `package.json` by Github.

File: typings/artifacts.d.ts [hunk 17]
```diff
@@ -53,6 +53,8 @@ declare global {
       CrawlableLinks: {href: string, text: string}[];
       /** CSS coverage information for styles used by page's final state. */
       CSSUsage: {rules: Crdp.CSS.RuleUsage[], stylesheets: Artifacts.CSSStyleSheetInfo[]};
+      /** Information on the document's Doctype, specifically the name, publicId, and systemId */
```
Comment (brendankenny) [inline typings/artifacts.d.ts]: maybe also something like "Properties will be the empty string if not present on the doctype." or something? I was expecting `undefined` or something but at least the latter two are specced to be the empty string if not present

File: typings/artifacts.d.ts [hunk 18]
```diff
@@ -53,6 +53,8 @@ declare global {
...
+      Doctype: Artifacts.Doctype;
```
Comment (brendankenny) [inline typings/artifacts.d.ts]: `Doctype: Artifacts.Doctype | null;`

File: lighthouse-core/gather/gatherers/dobetterweb/doctype.js [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+
+const Gatherer = require('../gatherer');
+
+/**
+ * Get and return `name`, `publicId`, `systemId` from
+ * `document.doctype`
+ * @return {Object<string>}
+ */
+function getDoctype() {
+  const {name, publicId, systemId} = document.doctype; // eslint-disable-line no-undef
```
Comment (brendankenny) [inline lighthouse-core/gather/gatherers/dobetterweb/doctype.js]: this will throw if there's no doctype, so I guess have an early return as well?
```js
if (!document.doctype) {
  return null;
}
...
```

Comment (schalkneethling) [reply lighthouse-core/gather/gatherers/dobetterweb/doctype.js]: much better, thx

File: lighthouse-core/audits/dobetterweb/doctype.js (hunk 5 above)
Comment (schalkneethling) [reply lighthouse-core/audits/dobetterweb/doctype.js]: I decided to add your suggestion in the end ;)

**Comment by schalkneethling** [comment]
Thank you for the review everyone. Updated.

**Comment by schalkneethling** [comment]
Oh I see one failing test when running `yarn test` but seems unrelated:

screen shot 2018-06-17 at 19 54 09

File: lighthouse-core/audits/dobetterweb/doctype.js (hunk 15 above)
Comment (schalkneethling) [reply lighthouse-core/audits/dobetterweb/doctype.js]: @paulirish You want me to add checks for those additional properties in this audit? Thanks!

Comment (paulirish) [reply lighthouse-core/audits/dobetterweb/doctype.js]: yup!

this applies to HTML pages as well. just make assert those params are empty string.

**Comment by schalkneethling** [comment]
@paulirish @brendankenny Ok, I added tests for `systemId` and `publicId`.

I am seeing a new TS error now:

```
lighthouse-core/gather/gatherers/dobetterweb/doctype.js(17,5): error TS2322: Type 'undefined' is not assignable to type '{ name: string; publicId: string; systemId: string; }'.
```

I am guessing it is happening because the function could return undefined here:
https://github.com/GoogleChrome/lighthouse/pull/5274/files#diff-5bd8ddd9e30281c7e7ec5103a145d002R17

**Comment by paulirish** [comment]
```diff
 /**
  * Get and return `name`, `publicId`, `systemId` from
  * `document.doctype`
- * @return {{name: string, publicId: string, systemId: string}}
+ * @return {{name: string, publicId: string, systemId: string}=}
  */
 function getDoctype() {
   if (!document.doctype) { // eslint-disable-line no-undef
```

**Comment by schalkneethling** [comment]
Also resolved the conflict, let me know and I should rebase if required.

**Comment by schalkneethling** [comment]
@paulirish @brendankenny Are there still some items outstanding in the PR? Thanks!

**Comment by brendankenny** [comment]
Not sure why appveyor keeps failing the smoke tests for this PR but Travis does not. We'll have to investigate.

**Comment by brendankenny** [comment]
Ah, I thought I saw a different failure before, but the failure now is also on master. Could be a Canary (appveyor, local) vs Stable (travis) issue?

File: lighthouse-core/audits/dobetterweb/doctype.js [hunk 20]
```diff
@@ -0,0 +1,80 @@
...
+  static audit(artifacts) {
+    if (!artifacts.Doctype) {
+      return {
+        rawValue: false,
+        explanation: 'Document must contain a doctype',
+      };
+    }
+
+    // only set constants once we know there is a doctype
+    const doctypeName = artifacts.Doctype ? artifacts.Doctype.name.trim() : '';
```
[the remark this answers is not in the document]
Comment (schalkneethling) [reply lighthouse-core/audits/dobetterweb/doctype.js]: That makes total sense. If it passed the check above, `artifacts.Doctype` should at the very least exist.

[the remark this answers is not in the document]
Comment (schalkneethling) [reply lighthouse-core/audits/dobetterweb/doctype.js]: Yup, following convention from the other audits.

**Comment by schalkneethling** [comment]
@paulirish @brendankenny updated. Apologies for how long this has taken.

File: lighthouse-core/test/audits/dobetterweb/doctype-test.js [hunk 21]
```diff
@@ -0,0 +1,92 @@
+/**
+ * @license Copyright 2017 Google Inc. All Rights Reserved.
```
Comment (ev1stensberg) [inline lighthouse-core/test/audits/dobetterweb/doctype-test.js]: 2018

File: lighthouse-core/gather/gatherers/dobetterweb/doctype.js [hunk 22]
```diff
@@ -0,0 +1,36 @@
+/**
+ * @license Copyright 2017 Google Inc. All Rights Reserved.
```
Comment (ev1stensberg) [inline lighthouse-core/gather/gatherers/dobetterweb/doctype.js]: 2018

File: lighthouse-core/audits/dobetterweb/doctype.js [hunk 23]
```diff
@@ -0,0 +1,73 @@
+/**
+ * @license Copyright 2017 Google Inc. All Rights Reserved.
```
Comment (ev1stensberg) [inline lighthouse-core/audits/dobetterweb/doctype.js]: 2018

File: lighthouse-core/test/results/sample_v2.json [hunk 24]
```diff
@@ -3287,7 +3300,7 @@
         }
       ],
       "id": "best-practices",
-      "score": 0
+      "score": null
```
Comment (ev1stensberg) [inline lighthouse-core/test/results/sample_v2.json]: Why is this null?

Comment (schalkneethling) [reply lighthouse-core/test/results/sample_v2.json]: I believe this file is generated by `update:sample-json` so, not something I specifically did in this PR. Perhaps I am wrong? @paulirish @brendankenny

Comment (patrickhulce) [reply lighthouse-core/test/results/sample_v2.json]: That's right @schalkneethling, but it's null because your audit is erroring when run on our fixtures 
 because you added a new artifact. Not ideal that we're adding an erroring audit to our golden, do you think you could manually add an expected Doctype artifact to https://github.com/GoogleChrome/lighthouse/blob/master/lighthouse-core/test/results/artifacts/artifacts.json?

Comment (schalkneethling) [reply lighthouse-core/test/results/sample_v2.json]: Thank you for the feedback @patrickhulce. I am happy to add it but, I probably need some guidance. Thanks!

**Comment by paulirish** [comment]
Open that artifacts.json up and add a Doctype object in there with 3
properties in it: name systemId publicId. With values we'd consider to be
passing.

**Comment by schalkneethling** [comment]
@paulirish Updates. I believe I did it correctly. The value is no longer `null`

**Comment by schalkneethling** [comment]
Fixed the failing tests. Let's see if we can get an all green build 😃

**Comment by paulirish** [comment]
Looks great! 

Nice job @schalkneethling and thanks for working with us on this one. :)

We're happy to have your contribution and happy to have this audit.
